### DISCUSSION TITLE
How to use type hints with constrained types

### DISCUSSION THREAD

**Discussion by IDrumsey** [pydantic/pydantic#5528] [Question] [answered]
When I use a constrained type like `conlist`, python shows the type of the variable as `Any`.

For example,

```
from pydantic import BaseModel, conlist



class TestClass(BaseModel):

    testList: conlist(num, min_items=1)
```

the TestClass.testList type annotation shows `Any` instead of `list[int]`. Is there a way to fix this, or is this just a flaw in the current system?

**Comment by dmontagu** [comment, marked as answer]
It's a flaw in the current system, but has been addressed in v2 where you use `Annotated` to add constraints [like this](https://github.com/pydantic/pydantic/blob/043e80b99e9500f1bca73fce655b711307efe68d/tests/test_model_signature.py#L165).

There are some tricks to get _better_ behavior in v1, such as doing something like:
```python
if typing.TYPE_CHECKING:
    NonEmptyListInt = list[int]
else:
    NonEmptyListInt = conlist(num, min_items=1)

class TestClass(BaseModel):

    testList: NonEmptyListInt
```
but we recognize this is not great, and so added improvements in v2.

**Reply by IDrumsey** [reply]
Great, thanks! Do you know generally when v2 is going to be out as a stable release? I saw the plan said end of Q1, but looks like it got pushed back.
